src/core/crypto.c

```diff
@@ -485,15 +485,18 @@ QuicCryptoOnVersionChange(
 _IRQL_requires_max_(PASSIVE_LEVEL)
 void
 QuicCryptoHandshakeConfirmed(
-    _In_ QUIC_CRYPTO* Crypto
+    _In_ QUIC_CRYPTO* Crypto,
+    _In_ BOOLEAN SignalBinding
     )
 {
     QUIC_CONNECTION* Connection = QuicCryptoGetConnection(Crypto);
     Connection->State.HandshakeConfirmed = TRUE;
 
-    QUIC_PATH* Path = &Connection->Paths[0];
-    CXPLAT_DBG_ASSERT(Path->Binding != NULL);
-    QuicBindingOnConnectionHandshakeConfirmed(Path->Binding, Connection);
+    if (SignalBinding) {
+        QUIC_PATH* Path = &Connection->Paths[0];
+        CXPLAT_DBG_ASSERT(Path->Binding != NULL);
+        QuicBindingOnConnectionHandshakeConfirmed(Path->Binding, Connection);
+    }
 
     QuicCryptoDiscardKeys(Crypto, QUIC_PACKET_KEY_HANDSHAKE);
```

Comment:
Does it make sense to throw away the handshake key still? Or should this whole function just be delayed?

Reply:
I don't think it makes sense to keep them, when TLS handshake finishes at server side, client already must have 1-RTT keys. The issue this fixes is only about getting the coalesced datagram routed to the connection.